src/mmap_xen.rs

```diff
@@ -0,0 +1,607 @@
+        }
+    }
+}
+
+impl MmapXenTrait for MmapXenForeign {
+    fn mmap_in_advance(&self) -> bool {
+        true
+    }
+
+    // Maps the entire guest memory space at once.
+    fn mmap(
+        &mut self,
+        size: usize,
+        prot: i32,
+        flags: i32,
+        file_offset: &Option<FileOffset>,
+    ) -> Result<*mut u8> {
+        let (fd, f_off) = validate_file(file_offset)?;
+        let (count, size) = pages(size);
+        let addr = mmap_unix::mmap_unix(size, prot, flags | MAP_SHARED, fd, f_off)?;
+        self.fd = fd;
+        self.ioctl(count, addr)?;
+
+        self.addr = addr;
+        self.size = size;
+
+        Ok(self.addr)
+    }
+
+    // Unmaps the memory.
+    fn munmap(&self) {
+        // SAFETY: This is safe because we mapped the addr earlier.
+        unsafe {
+            libc::munmap(self.addr as *mut c_void, self.size);
+        }
+    }
+
+    #[allow(unused_variables)]
+    fn mmap_slice(&self, addr: *const u8, is_read: bool, len: usize) -> Result<MmapXenSlice> {
+        panic!();
+    }
+}
+
+// Xen Grant memory mapping interface.
+
+const XEN_GRANT_ADDR_OFF: u64 = 1 << 63;
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+// Grant reference
+//
+// include/uapi/xen/gntdev.h: `ioctl_gntdev_grant_ref`
+struct GntDevGrantRef {
+    // The domain ID of the grant to be mapped.
+    domid: u32,
+    // The grant reference of the grant to be mapped.
+    r#ef: u32,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+// Grant dev mapping reference
+//
+// include/uapi/xen/gntdev.h: `ioctl_gntdev_map_grant_ref`
+struct GntDevMapGrantRef {
+    // The number of grants to be mapped.
+    count: u32,
+    // Unused padding
+    pad: u32,
+    // The offset to be used on a subsequent call to mmap().
+    index: u64,
+    // Array of grant references, of size @count.
+    refs: [GntDevGrantRef; 0],
+}
+
+impl GntDevMapGrantRef {
+    fn layout(count: usize) -> Layout {
+        let size = size_of::<Self>() + count * size_of::<GntDevGrantRef>();
+        Layout::from_size_align(size, align_of::<Self>()).unwrap()
+    }
+
+    fn new(domid: u32, base: u32, count: usize) -> &'static mut Self {
+        let layout = Self::layout(count);
+
+        // SAFETY: This is safe as we are allocating a fixed size of memory that needs to be passed
+        // to the ioctl later.
+        let (map, refs) = unsafe {
+            let ptr = alloc::alloc(layout);
+            if ptr.is_null() {
+                handle_alloc_error(layout);
+            }
+
+            let map: &mut Self = &mut *(ptr as *mut Self);
+            let refs: &mut [GntDevGrantRef] =
+                slice::from_raw_parts_mut(ptr.add(size_of::<Self>()) as *mut GntDevGrantRef, count);
+
+            (map, refs)
+        };
+
+        map.count = count as u32;
+        map.pad = 0;
+        map.index = 0;
+
+        for (i, r) in refs.iter_mut().enumerate().take(count) {
+            r.domid = domid;
+            r.r#ef = base + i as u32;
```

Comment:
Where does this `#` come from in `r#ef`?

Reply:
`ref` is a keyword, had to do this to use the same name. Named it `reference` now.